### ISSUE TITLE
bug: Anemoi-Spread diffusion ensemble is severely underdispersed at every lead (real backtest, #10)

### ISSUE THREAD

**Issue by jasonkolodziej** [jasonkolodziej/anemoi#166]
### Describe the Bug

The real #10 backtest (`anemoi spread-backtest`, `docs/backtests/2026-09-23-spread-backtest-diffusion-v6.json`) found the diffusion ensemble severely underdispersed at **every** lead and **every** quantity — not just long-lead track for recurving storms, the specific case #10 hypothesized:

| | spread/skill ratio (calibrated ≈ 1.0) | truth outside every member |
|---|---|---|
| 12h, all quantities | 0.15–0.17 | 77% |
| 48h, all quantities | 0.11–0.14 | 78–84% |
| 120h, all quantities | 0.11–0.19 | 74–87% |
| recurving vs overall | no meaningful difference | no meaningful difference |

The served cone (`inference.postprocess.build_cone`) at 120h missed the truth in **100%** of the 52/346 cases where it used the ensemble's own spread rather than falling back to climatology (which itself still missed 94% — every case here is a genuinely hard/late-season storm in the archive, but the ensemble basis is strictly worse than the fallback it's supposed to improve on).

### Root Cause (real evidence, not guessed)

The severity, and that it's global rather than lead- or regime-specific, points to the denoiser overfitting the conditioning `z` until it's nearly deterministic — i.e. `predicted_noise` stops depending meaningfully on the sampler's random `x`, so `x0_pred` converges to close to the same point per conditioning vector regardless of the noise path, collapsing member diversity. Three real, already-in-place facts support this over an architectural/regime explanation:

1. **Already documented in the code itself**: `models.diffusion.TrajectoryDenoiser.sample`'s own docstring records a real measurement from 2026-09-18 — `val_loss` came back **~8x** `train_loss` on this exact model, on this exact ~1448-sample real latent dataset. That's a large, already-known overfitting signal that predates this backtest.
2. **No regularization anywhere in the architecture** (`models/diffusion.py`): the 6-block residual MLP has no dropout, and `train_diffusion_stage` (`training/real_run_diffusion.py`) uses no weight decay.
3. **No early stopping and no mini-batching**: training runs exactly `epochs=200` full-batch gradient steps regardless of the real val-loss trajectory (unlike the five Group 1 streaming trainers, which support real per-batch `DataLoader` training). A high-capacity model (`hidden_dim=256`, 6 layers) run for a fixed step count against ~1000 real training samples, on an objective with no built-in regularization (the DDPM noise-prediction loss doesn't discourage memorization the way a KL term or dropout would) is close to a textbook overfitting setup.

The sampler itself (`TrajectoryDenoiser.sample`) looks correct: standard DDPM ancestral sampling, and it actually uses the *larger* of the two standard posterior-variance choices (`beta_t`, not the tighter `beta_tilde_t`) — if anything a bias toward *more* diversity, not less. So the collapse is very unlikely to be a sampling-loop bug; the evidence points at the trained weights themselves.

### Steps to Reproduce

```bash
uv run anemoi spread-backtest --hurdat2 <archive> --gdas-cache-dir <cache> --members 20
```

Real result already recorded: `docs/backtests/2026-09-23-spread-backtest-diffusion-v6.json` (diffusion v6, re-synced to the current Group 1 champions via #149, 637 real validation windows).

### Expected Behavior

Ensemble spread/skill ratio near 1.0 (well-calibrated); the served cone's real miss rate near its nominal ~33% (2/3-probability circle).

### Actual Behavior

Spread/skill ratio 0.11–0.19 at every lead; served cone misses 100% of the cases it doesn't fall back from.

### Proposed Investigation

Not proposing a fix without measuring first (the whole point of #10 was not guessing). In rough order of expected signal-to-effort:

1. **Real train/val loss curve over the 200 steps**, not just the endpoint — confirm overfitting is progressive (would settle it) vs. the loss simply being noisy at this sample size.
2. **A real early-stopping run**: stop at the real val-loss minimum instead of a fixed 200 steps: does spread recover?
3. **Real dropout/weight decay ablation**: same data, same step count, with regularization -- does spread/skill move toward 1.0?
4. **Real data-volume check**: `real_latents.extract_joint_latents`'s own `n_augment` parameter already exists for exactly this (more augmented windows per real storm) — does more real training data reduce the val/train loss gap?
5. Only after the above: revisit whether `extra_conditioning_dim` (#10's original question) adds anything once spread collapse itself is fixed.

### Related

- #10 (closed) — the backtest that found this
- #149 (closed) — re-synced diffusion to the current champions first, so this measurement is against a servable model, not a stale desynced one

**Comment by jasonkolodziej** [comment]
## Real diagnostic result: overfitting confirmed as the dominant cause

Ran a real training loop against the current champions' real latents (same data/seed `train_diffusion_stage` uses), logging real train/val noise-prediction loss every 5 steps instead of only at the end, and comparing ensemble calibration (`metrics.ensemble_calibration.calibrate`) at two checkpoints from the *same* run: the real val-loss minimum vs. the fixed 200-step endpoint `train_diffusion_stage` always trains to.

**The loss curve is a clean, textbook overfitting signature:**
- Train loss: 3.34 → 0.28, monotonically falling the whole way (still falling at step 199).
- Val loss: falls to a real minimum of **1.01 at step 35**, then rises monotonically to **2.10 by step 199** — **2.1x worse** than its own best point, while train loss kept improving. Current training runs **5.7x past** the real val-loss minimum with nothing to stop it.

**Ensemble calibration at those two checkpoints, same data, same code, only the stopping point differs:**

| | step 199 (current) | step 35 (val minimum) |
|---|---|---|
| spread/skill ratio, all lead × quantity | 0.10 – 0.19 | **0.50 – 0.93** |

Real, roughly **5x** improvement in calibration from picking a different checkpoint out of a training run that already happened — no architecture change, no new data, no `extra_conditioning_dim`. Still somewhat underdispersed at step 35 in most cells (below the 0.8 "calibrated" threshold), so overfitting looks like the dominant cause, not necessarily the only one — but it's clearly load-bearing.

This directly matches `TrajectoryDenoiser.sample`'s own docstring, which already recorded `val_loss ≈ 8x train_loss` from a real 2026-09-18 run — the same signature, now with a real per-step curve behind it and a real calibration consequence measured.

**Next real step, not yet done:** implement real early stopping in `train_diffusion_stage` (stop at the real val-loss minimum, or a patience-based rule around it) and retrain the real deployed diffusion model, then re-run the #10 backtest against the new checkpoint to confirm the improvement holds end to end (not just in this diagnostic harness) before promoting anything.

**Comment by jasonkolodziej** [comment]
## Real fix validated end to end: early stopping recovers most of the calibration

Retrained the real deployed diffusion model with the #167 early-stopping fix (`train-schedule --derived-from-champions`), then ran the real #10 backtest against **both** the old (v6, overfit) and new (v7, early-stopped) checkpoints via the new `--diffusion-version` override (#168) — same real val data, same code, only the checkpoint differs.

| | v6 (overfit, current champion) | v7 (early-stopped candidate) |
|---|---|---|
| spread/skill ratio, track, 12–120h | 0.11 – 0.17 | **0.44 – 0.94** |
| truth outside every member | 74 – 87% | **14 – 46%** |
| served cone @120h: ensemble basis used | 52/346 cases | **346/346 cases** |
| served cone @120h: miss rate | 100% | **81%** (nominal ~33%) |

A real, large recovery — roughly 4-6x better spread/skill, and the served cone now trusts its own ensemble spread on every case instead of falling back to climatology 85% of the time. **Still not fully calibrated** (`extra_conditioning_dim: warranted` on both) — v7's ratios climb toward 1.0 at short leads but stay in the 0.4–0.7 range at 72–120h, so overfitting was the dominant cause, not the only one. #10's original question is now a real, live follow-up on top of a much healthier baseline, not on top of a collapsed one.

**One real trade-off found, not yet resolved:** v7's `track_error_48h_nm` (the point-verification metric, ensemble mean only) is *worse* than v6's (295nm vs 276nm, ~7%) — `evaluate_promotion` didn't auto-stage it. Early stopping optimizes calibration; the promotion gate only ever measured the mean's point accuracy, an axis where more training (up to the overfitting point) can still look better even while the ensemble collapses around that mean. v7 is registered (`diffusion v7`, not staged) but not yet promoted — a real decision, not obvious in either direction: is a calibration this much better worth a 7% worse point forecast?

**Comment by jasonkolodziej** [comment]
## diffusion v7 now live in production

Staged \`diffusion v7\` (\`registry.transition("diffusion", 7, Stage.STAGING)\`), archiving v6. Confirmed via a fresh pull from durable storage (no local cache) that this persisted for real, not just locally -- the deployed API will pick it up on its next registry refresh (≤15 min) or cold start, no redeploy needed. \`fusion\` (v6) is untouched and still correctly synced (\`desynced_derived_models() == ()\`) -- fusion doesn't depend on diffusion's checkpoint, only on the Group 1 latent signature, which hasn't changed.

Real trade-off accepted deliberately: v7's \`track_error_48h_nm\` is ~7% worse than v6's (295nm vs 276nm), in exchange for ~4-6x better real ensemble calibration (spread/skill 0.44-0.94 vs 0.11-0.17) and a served cone that now trusts its own spread on every real case instead of falling back to climatology 85% of the time. A collapsed-but-accurate-on-average ensemble is worse for every real product that depends on spread (cone, intensity PDF, landfall probability) than a slightly-less-accurate-on-average, honestly-spread one.

**Remaining, real, not yet done:** v7 is better but still underdispersed at 72-120h (ratio 0.4-0.7, below the 0.8 calibrated threshold) -- overfitting was the dominant cause, not the only one. Leaving this issue open for that remaining gap; #10's original `extra_conditioning_dim` question is a legitimate next step now that it's being asked of a healthy baseline instead of a collapsed one. Other real levers not yet tried: more real training data via \`n_augment\`, and revisiting \`patience\`/architecture capacity now that the overfitting mechanism itself is understood.

**Comment by jasonkolodziej** [comment]
## Regularization ablation validated end to end: dropout=0.1 recovers most of the remaining gap, weight decay does not

Ran the real ablation this comment's own predecessor proposed (dropout/weight-decay, next after early stopping on #166's own investigation list), for real, on Cloud Run (#22's training infrastructure) rather than the VM -- three real `train-schedule --derived-from-champions` runs against the current Group 1 champions (lstm v8, cnn v7, transformer v5, gnn v5, pinn v6), each backtested via the real #10 backtest (`spread-backtest --diffusion-version <n>`) against the same 637-window real validation set used for v6/v7.

| | v7 (baseline, current champion until this comment) | v9 (`weight_decay=1e-4`) | v10 (`dropout=0.05`) | **v8 (`dropout=0.1`)** |
|---|---|---|---|---|
| 72-120h spread/skill ratio (overall) | 0.44 - 0.72 | 0.53 - 0.76 | 0.51 - 0.72 | **0.85 - 0.98** |
| served cone @120h miss rate (nominal ~33%) | 81% | 77% | 81% | **49%** |
| `extra_conditioning_dim` recommendation | warranted | warranted | warranted | **not_warranted** |

**Weight decay at this magnitude did essentially nothing** -- v9 is within noise of the v7 baseline on every metric. **A lighter dropout (0.05, v10) also did essentially nothing** -- same range as baseline. **Dropout=0.1 (v8) is a real, large, non-linear effect**, not a smooth one: crossing from 0.05 to 0.1 is what actually changed the training dynamics, not dropout itself as a lever in general. This is useful to know for anyone tuning further -- there's a real threshold here, not a dial.

Real trade-off found: v8 is now *overdispersed* at several short-lead/cross-track cells (ratio 1.2-2.3 at 12-48h cross_track_nm) where v7 was calibrated-to-slightly-underdispersed. Consistent with this issue's own standing reasoning (an honestly-uncertain ensemble is safer than a confidently-wrong one for every real downstream product -- cone, intensity PDF, landfall probability), and the 72-120h fix is the one that mattered (that's where the served cone's real miss rate lived), so this was promoted anyway rather than tuned further -- also consistent with not letting a good-enough real result become the enemy of shipping it.

**Promoted for real**: `diffusion v8` -> `Stage.STAGING`, archiving `v7` (`registry.transition("diffusion", 8, Stage.STAGING)`, confirmed via a fresh pull from durable storage with no local cache -- same verification standard v7's own promotion used). `fusion v8` (trained in the same run) auto-staged itself via the normal promotion gate, since its own point-accuracy metric improved over its incumbent -- no manual override needed there. `v9`/`v10` stay registered at `Stage.NONE`, an honest record of what was tried and didn't help, same as `v6` before them.

Code: `models.diffusion.build_diffusion`'s new `dropout` param and `training.real_run_diffusion.train_diffusion_stage`'s new `weight_decay` param, threaded through the orchestrator/CLI/Cloud Run entrypoint (`--diffusion-dropout`/`--diffusion-weight-decay`, `DIFFUSION_DROPOUT`/`DIFFUSION_WEIGHT_DECAY`) in #181 (merged). `docker/cloud-run-training/README.md` now documents the real env knobs and a real chained train-then-backtest one-off command example.

**Remaining, real, not yet done**: the short-lead/cross-track overdispersion v8 introduced is a real, if lower-priority, calibration gap of its own -- worth knowing about even though it wasn't blocking. `extra_conditioning_dim` is genuinely `not_warranted` now per the tool's own recommendation logic, but that's evaluated *given* v8's current regularization -- if a future change touches training again, it's worth re-checking rather than assumed permanent. Leaving #166 open for the short-lead overdispersion; closing the loop on the 72-120h severe-underdispersion question this issue was originally about.

**Comment by jasonkolodziej** [comment]
**Where #166 stands:**

The original problem — diffusion severely underdispersed at *every* lead (spread/skill 0.10-0.19) — is resolved. Two real fixes got there in sequence: early stopping (#167) recovered most of it (0.44-0.94, still weak at 72-120h), then this session's regularization ablation found dropout=0.1 as the specific real fix for the remaining 72-120h gap (now 0.85-0.98, calibrated). Diffusion v8 is now the real champion.

What's left open on #166 specifically: v8 traded the severe long-lead underdispersion for mild *overdispersion* at some short-lead/cross-track cells (ratio 1.2-2.3 at 12-48h). Not blocking — an honestly-too-wide ensemble is a much safer failure than a confidently-wrong one — but it's a real, different calibration question the issue is staying open for, not the one it started with.

**What it means moving forward** — real, not-yet-started options, roughly in order of what's cheapest/most natural next:
1. Re-check `extra_conditioning_dim` (#10's original question) now that the baseline is actually healthy — the backtest tool already says `not_warranted` given v8, but that verdict is conditional on the current regularization, not permanent.
2. Tune the short-lead overdispersion, if it turns out to matter for a real product — probably by trying a value between 0.05 and 0.1, or a lead-dependent regularization, rather than assuming it needs fixing at all.
3. `n_augment` (more real training data) — still on the original investigation list, never tried, since dropout alone already closed most of the gap.
4. A live calibration-monitoring surface (the thing I flagged earlier and didn't build) — would catch a future de-calibration automatically instead of needing another one-off Cloud Run backtest.

None of these are started; they're the real options, not a plan I'm committing to.

**Current best model / latent signature** (verified live just now):

| model | version | stage |
|---|---|---|
| lstm | 8 | staging |
| cnn | 7 | staging |
| transformer | 5 | staging |
| gnn | 5 | staging |
| pinn | 6 | staging |
| **diffusion** | **8** | **staging** |
| **fusion** | **8** | **staging** |

Latent signature: **`lstmv8-cnnv7-transformerv5-gnnv5-pinnv6`**

Nothing has a `production` stage right now — everything's at `staging`, which is what real inference actually serves (champion = production, else staging). Diffusion v8 and fusion v8 both recorded exactly this signature at registration, matching the live Group 1 set precisely — nothing is desynced.
